### ISSUE TITLE
Overloading error w/ matrix-matrix multiply

### ISSUE THREAD

**Issue by dsimcha** [cristicbz/scid#20]
```
import scid.matvec, std.stdio;

void main(string[] args) {
    auto x = Matrix!double([[1.0, 4, 3], [4.0, 5, 6], [7.0, 8, 9]]);
    auto v = Vector!double([2.0, 8, 7]);
    auto expr = x * x;
}
```

C:\code\test.d(6): Error: overloads nothrow @system Expression!(op,BasicMatrix!(BasicGeneralMatrixStorage!(RefCounted!(CowMatrix!(double),cast(RefCountedAutoInitialize)1))),BasicMatrix!(BasicGeneralMatrixStorage!(RefCounted!(CowMatrix!(double),cast(RefCountedAutoInitialize)1))))(auto ref BasicMatrix!(BasicGeneralMatrixStorage!(RefCounted!(CowMatrix!(double),cast(RefCountedAutoInitialize)1))) newRhs) and nothrow @system Expression!(op,BasicMatrix!(BasicGeneralMatrixStorage!(RefCounted!(CowMatrix!(double),cast(RefCountedAutoInitialize)1))),BasicMatrix!(BasicGeneralMatrixStorage!(RefCounted!(CowMatrix!(double),cast(RefCountedAutoInitialize)1))))(BasicMatrix!(BasicGeneralMatrixStorage!(RefCounted!(CowMatrix!(double),cast(RefCountedAutoInitialize)1))) newLhs) both match argument list for opBinary

**Comment by cristicbz** [comment]
I can't reproduce this, nor the other one, could you try getting a fresh copy of the repo?

**Comment by dsimcha** [comment]
I've double and triple checked everything and it still doesn't work.  Maybe
you forgot to commit something?

**Comment by cristicbz** [comment]
Nope, I had created a new folder and pulled from the repo to try to reproduce it and it still works. Maybe it's the compiler? I'm using a slightly modified phobos & druntime,but the compiler itself is the one that comes with the std D 2.054 installation. Should I grab the HEAD version?
What seems to happen is that for some reason opBinaryRight gets called when it shouldn't. I'm using a trait called "isConvertible!(S,T)" to check whether to!Something works for a given type.
Could you please check if isConvertible!(Matrix!double, double) returns true at your end?

**Comment by dsimcha** [comment]
Yes, isConvertible!(Matrix!double, double) returns true at my end.  I assume
it shouldn't.  I did have a few minor modifications to Phobos as well (I
don't remember what they were) so I went and downloaded a 2.054 stock
install.  It made no difference.  All I can figure is that one of those
small modifications somehow makes isConvertible work properly.

**Comment by cristicbz** [comment]
Ok, so this is truly weird. I got myself a fresh install as well and on mine it still works....
I commited a different version of isConvertible, could you check if it works with the new one?

**Comment by dsimcha** [comment]
I played around with this a little.  It definitely looks like a compiler bug is at the root of the problem.  Even though isConvertible!(Matrix!double, double) returns true, to!(double)(Matrix!(double).init) doesn't actually compile.  As for why you're not affected, I haven't the slightest clue.  

Here's an implementation of isConvertible that seems to work around the compiler bug.  I came up with this by pure trial and error.  Using this implementation also fixes Issue 21, which you were also not able to reproduce.

```
 template isConvertible( S, T ) {
     enum isConvertible = is( typeof({
         return to!(T)(S.init);
     }));
 }
```
